### PR TITLE
SDK::Trace::Span defers attr validation until finish is called.

### PR DESCRIPTION
**Pull request by fhwang**

### CODE REVIEW

**Comment by fhwang** [comment]
@mwear @fbogsany This related to our conversation on #203. The primary intention is to centralize validation around attributes in preparation for adding more attribute value validation.

File: sdk/lib/opentelemetry/sdk/trace/span.rb [hunk 1]
```diff
@@ -203,8 +202,13 @@ def finish(end_timestamp: nil)
               return self
             end
             @end_timestamp = end_timestamp || Time.now
-            @attributes.freeze
-            @events.freeze
+            if @attributes
+              @attributes = finished_component_attributes(
```
Comment (mwear) [inline sdk/lib/opentelemetry/sdk/trace/span.rb]:
```suggestion
              @attributes = validated_attributes(
```

File: sdk/lib/opentelemetry/sdk/trace/span.rb [hunk 2]
```diff
@@ -203,8 +202,13 @@ def finish(end_timestamp: nil)
...
+                @attributes, @trace_config.max_attributes_count
+              )
+            end
+            @events = finished_events_for_processor
```
Comment (mwear) [inline sdk/lib/opentelemetry/sdk/trace/span.rb]:
```suggestion
            @events = validated_events
```

File: sdk/lib/opentelemetry/sdk/trace/span.rb [hunk 3]
```diff
@@ -203,8 +202,13 @@ def finish(end_timestamp: nil)
...
+            @links = finished_links_for_processor
```
Comment (mwear) [inline sdk/lib/opentelemetry/sdk/trace/span.rb]:
```suggestion
            @links = validated_links
```

File: sdk/lib/opentelemetry/sdk/trace/span.rb [hunk 4]
```diff
@@ -262,69 +266,59 @@ def initialize(context, name, kind, parent_span_id, trace_config, span_processor
...
-        def append_event(events, event) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
-          max_events_count = @trace_config.max_events_count
-          max_attributes_per_event = @trace_config.max_attributes_per_event
-
-          # Fast path (likely) common case.
-          if events.size < max_events_count &&
-             event.attributes.size <= max_attributes_per_event &&
-             Internal.valid_attributes?(event.attributes)
-            return events << event
+        def finished_component_attributes(original_attrs, max_attrs_count)
```
Comment (mwear) [inline sdk/lib/opentelemetry/sdk/trace/span.rb]:
```suggestion
        def validated_attributes(original_attrs, max_attrs_count)
```

File: sdk/lib/opentelemetry/sdk/trace/span.rb [hunk 5]
```diff
@@ -262,69 +266,59 @@ def initialize(context, name, kind, parent_span_id, trace_config, span_processor
...
           end
+          excess = attrs.size - max_attrs_count
+          excess.times { attrs.shift } if excess.positive?
+          attrs.freeze
+        end
 
-          # Slow path.
-          excess = events.size + 1 - max_events_count
-          events.shift(excess) if excess.positive?
+        def finished_events_for_processor
```
Comment (mwear) [inline sdk/lib/opentelemetry/sdk/trace/span.rb]:
```suggestion
        def validated_events
```

File: sdk/lib/opentelemetry/sdk/trace/span.rb [hunk 6]
```diff
@@ -262,69 +266,59 @@ def initialize(context, name, kind, parent_span_id, trace_config, span_processor
...
-
-          excess = attrs.size - @trace_config.max_attributes_count
-          # TODO: with Ruby 2.5, replace with the more efficient
-          # attrs.shift(excess) if excess.positive?
-          excess.times { attrs.shift } if excess.positive?
-          nil
-        end
-
-        def trim_links(links, max_links_count, max_attributes_per_link) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+        def finished_links_for_processor # rubocop:disable Metrics/AbcSize
```
Comment (mwear) [inline sdk/lib/opentelemetry/sdk/trace/span.rb]:
```suggestion
        def validated_links # rubocop:disable Metrics/AbcSize
```

**Comment by mwear** [comment]
@fhwang Are you interested in finishing this?

**Comment by fhwang** [comment]
@mwear Sorry about dropping off there. I've had a bunch of other things come up so for the foreseeable future I won't be a very reliable contributor to Otel-Ruby. If someone else wanted to take this that would make sense for now.

**Comment by mwear** [comment]
No problem and thanks for your help @fhwang! We'll put this up for grabs.

**Comment by fhwang** [comment]
Hey guys, I think I'll be able to do some more OTel-Ruby for the near future -- should I revive this dormant PR? Or is there something else that would be better for me to work on?

**Comment by fbogsany** [comment]
I believe this is replaced by #521

**Comment by fbogsany** [comment]
> Hey guys, I think I'll be able to do some more OTel-Ruby for the near future -- should I revive this dormant PR? Or is there something else that would be better for me to work on?

@fhwang sorry, both @mwear and I meant to respond much more promptly. I think we've covered this in #521. We'd really appreciate any contributions you can make. Our particular focus at the moment is [spec compliance](https://github.com/open-telemetry/opentelemetry-ruby/issues?q=is%3Aissue+is%3Aopen+label%3Aspec-compliance), though additional instrumentation, the Jaeger propagator or a Zipkin exporter would be welcome too.

**Comment by fhwang** [comment]
Cool, moving convo to Gitter.

**Pull request closed without merging**